### ISSUE TITLE
Crashes when calling date_trunc with certain values

### ISSUE THREAD

**Issue by jacobobryant** [MaterializeInc/materialize#3651]
### What version of Materialize are you using?

```
$ materialized -v
materialized v0.3.1 (9e4dce4bee6169e8391d9d7a6f5382b998a00625)
```

### How did you install Materialize?

* [ ] Docker image
* [x] Linux release tarball
* [ ] APT package
* [ ] macOS release tarball
* [ ] Homebrew tap
* [ ] Built from source

### What was the issue?

`date_trunc` crashes materialized for some dates.

### Is the issue reproducible? If so, please provide reproduction instructions.

```
materialized --w=1 &
psql -h localhost -p 6875 materialize -c "select date_trunc('week', '2020-07-05 19:58:52+00'::timestamptz)"
```

### Please attach any applicable log files.

[materialized.log](https://github.com/MaterializeInc/materialize/files/4941277/materialized.log)

**Comment by ruchirK** [comment]
Confirmed that this is indeed a crash with the current version of materialize -- basically when trying to compute the truncated date we subtract the number of days since the most recent Monday, but we fail to appropriately go back to earlier months when needed

So something like

```
select date_trunc('week', '2020-08-02 00:00:00+00'::timestamptz);
```

August 2nd is a Sunday this year, so it's 6 days since the last Monday, and (2 - 6) underflows

**Comment by ruchirK** [comment]
@jacobobryant thanks for this bug report! I have a commit that fixes this issue in #3656 - should get merged in tomorrow

**Comment by jacobobryant** [comment]
Thanks!
